opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,323 @@
+    const INITIAL_VEC_CAPACITY: usize = DEFAULT_INITIAL_VEC_CAPACITY,
+> {
+    initial: [T; MAX_STACK_CAPACITY],
+    additional: Option<Vec<T>>,
+    count: usize,
+}
+
+impl<
+        T: Default + Clone + PartialEq,
+        const MAX_STACK_CAPACITY: usize,
+        const INITIAL_VEC_CAPACITY: usize,
+    > Default for GrowableArray<T, MAX_STACK_CAPACITY, INITIAL_VEC_CAPACITY>
+{
+    fn default() -> Self {
+        Self {
+            initial: [(); MAX_STACK_CAPACITY].map(|_| T::default()),
+            additional: None,
+            count: 0,
+        }
+    }
+}
+
+impl<
+        T: Default + Clone + PartialEq,
+        const MAX_STACK_CAPACITY: usize,
+        const INITIAL_VEC_CAPACITY: usize,
+    > GrowableArray<T, MAX_STACK_CAPACITY, INITIAL_VEC_CAPACITY>
+{
+    /// Creates a new `GrowableArray` with the default initial capacity.
+    #[allow(dead_code)]
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    /// Pushes a value into the `GrowableArray`.
+    ///
+    /// If the internal array (`initial`) has reached its capacity (`MAX_STACK_CAPACITY`), the value is pushed
+    /// into the heap-allocated vector (`additional`). Otherwise, it is stored in the array.    #[inline]
+    pub(crate) fn push(&mut self, value: T) {
+        if self.count < MAX_STACK_CAPACITY {
+            self.initial[self.count] = value;
+            self.count += 1;
+        } else {
+            if self.additional.is_none() {
+                // Initialize the vector with a specified capacity
+                self.additional = Some(Vec::with_capacity(INITIAL_VEC_CAPACITY));
+            }
+            self.additional.as_mut().unwrap().push(value);
+        }
+    }
+
+    /// Gets a reference to the value at the specified index.
+    ///
+    /// Returns `None` if the index is out of bounds.
+    #[allow(dead_code)]
+    pub(crate) fn get(&self, index: usize) -> Option<&T> {
+        if index < self.count {
+            Some(&self.initial[index])
+        } else if let Some(ref additional) = self.additional {
+            additional.get(index - MAX_STACK_CAPACITY)
+        } else {
+            None
+        }
+    }
+
+    /// Returns the number of elements in the `GrowableArray`.
+    pub(crate) fn len(&self) -> usize {
+        self.count + self.additional.as_ref().map_or(0, Vec::len)
+    }
+
+    /// Returns an iterator over the elements in the `GrowableArray`.
+    ///
+    /// The iterator yields elements from the internal array (`initial`) first, followed by elements
+    /// from the vector (`additional`) if present. This allows for efficient iteration over both
+    /// stack-allocated and heap-allocated portions.
+    ///
+    #[inline]
+    pub(crate) fn iter(&self) -> GrowableArrayIter<'_, T, MAX_STACK_CAPACITY> {
+        if self.additional.is_none() || self.additional.as_ref().unwrap().is_empty() {
+            GrowableArrayIter::StackOnly {
+                iter: self.initial.iter().take(self.count),
```

Comment:
Do we need to use `take`? It should have the same behavior as plain `iter()`, right?